WordPress/gutenberg
# Reusable blocks: Improve UX for non-privileged users

### What this is

Fixes #12338.

Improves the UX of creating, editing, and deleting a reusable block when logged in as an author or contributor by disabling the _Add to Reusable Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when necessary.

This is accomplished under the hood by introducing the `canUser()` selector to `core-data` which allows callers to query whether the REST API supports performing a given action on a given resource, e.g. one can query whether the logged in user can create posts by running `wp.data.select( 'core' ).canUser( 'create', 'posts' )`.

The existing `hasUploadPermissions()` selector is changed to use `canUser( 'create', 'media' )` under the hood.

### How it looks

_Add to Reusable Blocks_ button is visible for admins and editors:

![admin - add](https://user-images.githubusercontent.com/612155/50619911-71a5c380-0f50-11e9-8f30-acf8304b81e6.png)

_Add to Reusable Blocks_ button is hidden for contributors:

![contributor - add](https://user-images.githubusercontent.com/612155/50619915-78343b00-0f50-11e9-84c2-03dbee5f21b9.png)

_Edit_ button is enabled and _Remove from Reusable Blocks_ is visible for admins and editors:

![admin - edit remove](https://user-images.githubusercontent.com/612155/50619935-9601a000-0f50-11e9-8a6f-a679c9da4155.png)

_Edit_ button is disabled and _Remove from Reusable Blocks_ is hidden for authors:

![author - edit remove](https://user-images.githubusercontent.com/612155/50619918-86825700-0f50-11e9-969f-1c139bc113a4.png)

### How to test

1. Try to create a reusable block as a contributor. The _Add to Reusable Blocks_ button should be disabled.
1. Try to updating another user's reusable block as an author. The _Edit_ button should be disabled.
1. Try to deleting another user's reusable block as an author or contributor. The _Remove from Reusable Blocks_ button should be disabled.
